asserts/cluster.go

```diff
@@ -0,0 +1,323 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/snap/channel"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// Cluster holds a cluster assertion, which describes a cluster of devices and
+// their organization into subclusters.
+type Cluster struct {
+	assertionBase
+	seq         int
+	devices     []ClusterDevice
+	subclusters []ClusterSubcluster
+}
+
+// ClusterDevice holds the details about a device in a cluster assertion.
+type ClusterDevice struct {
+	// ID is the unique identifier referenced from subclusters.
+	ID int
+	// BrandID is the device's owning brand.
+	BrandID string
+	// Model is the device's snap model name.
+	Model string
+	// Serial is the device's serial identifier.
+	Serial string
+	// Addresses contains this device's known IP addresses.
+	Addresses []string
+}
+
+// ClusterSubcluster holds the details about a subcluster in a cluster
+// assertion.
+type ClusterSubcluster struct {
+	// Name is the subcluster's name.
+	Name string
+	// Devices lists device IDs that belong to this subcluster.
+	Devices []int
+	// Snaps contains the expected snap state for this subcluster.
+	Snaps []ClusterSnap
+}
+
+// ClusterSnap holds the details about a snap in a subcluster.
+type ClusterSnap struct {
+	// State describes the snap's state in the cluster (clustered, evacuated,
+	// removed).
+	State string
+	// Instance is the snap's instance name.
+	Instance string
+	// Channel is the channel the snap should track.
+	Channel string
+}
+
+// ClusterID returns the cluster's ID.
+func (c *Cluster) ClusterID() string {
+	return c.HeaderString("cluster-id")
+}
+
+// Sequence returns the sequence number of this cluster assertion.
+func (c *Cluster) Sequence() int {
+	return c.seq
+}
+
+// Devices returns the list of devices in the cluster.
+func (c *Cluster) Devices() []ClusterDevice {
+	return c.devices
+}
+
+// Subclusters returns the list of subclusters.
+func (c *Cluster) Subclusters() []ClusterSubcluster {
+	return c.subclusters
+}
+
+var validClusterSnapStates = []string{"clustered", "evacuated", "removed"}
```

Comment:
Since we're storing these states after checking them, should we keep them as a typed representation instead of just a string?